File: 19_数学/矩阵运算/矩阵快速幂/matqpow.ts
/* eslint-disable no-inner-declarations */

// 矩阵快速幂
// 长度较小时，Uint32Array的优势不明显，所以这里的二维矩阵直接用number[][]。
// 如果可以将矩阵压缩成一维数组，则使用Uint32Array优势更明显。

function newMatrix(row: number, col: number): number[][] {
  const res: number[][] = Array(row)
  for (let i = 0; i < row; i++) {
    res[i] = Array(col).fill(0)
  }
  return res
}

/**
 * 单位矩阵.
 * @example
 * ```ts
 * eye(3)
 * // [
 * //   [1, 0, 0],
 * //   [0, 1, 0],
 * //   [0, 0, 1]
 * // ]
 * ```
 */
function eye(n: number): number[][] {
  const res: number[][] = Array(n)
  for (let i = 0; i < n; i++) {
    const row = Array(n).fill(0)
    row[i] = 1
    res[i] = row
  }
  return res
}

function copy(raw: number[][]): number[][] {
  const row = raw.length
  const res: number[][] = Array(row)
  for (let i = 0; i < row; i++) {
    res[i] = raw[i].slice()
  }
  return res
}

/**
 * uint32乘法.
 */
function mulUint32(num1: number, num2: number, mod = 1e9 + 7): number {
  return (((Math.floor(num1 / 65536) * num2) % mod) * 65536 + (num1 & 65535) * num2) % mod
}

/**
 * uint32矩阵乘法.
 */
function matMul(mat1: number[][], mat2: number[][], mod = 1e9 + 7): number[][] {
  const row1 = mat1.length
  const row2 = mat2.length
  const col2 = mat2[0].length
  const res = newMatrix(row1, col2)
  for (let i = 0; i < row1; i++) {
    const resRow = res[i]
    const m1Row = mat1[i]
    for (let k = 0; k < row2; k++) {
      const m2Row = mat2[k]
      for (let j = 0; j < col2; j++) {
        resRow[j] = (resRow[j] + mulUint32(m1Row[k], m2Row[j], mod)) % mod
        if (resRow[j] < 0) resRow[j] += mod
      }
    }
  }
  return res
}

/**
 * uint32矩阵快速幂.
 */
function matPow(base: number[][], exp: number, mod = 1e9 + 7): number[][] {
  if (base.length !== base[0].length) throw new Error('base is not a square matrix')
  const n = base.length
  let res = eye(n)
  let baseCopy = copy(base)
  while (exp) {
    if (exp & 1) res = matMul(res, baseCopy, mod)
    baseCopy = matMul(baseCopy, baseCopy, mod)
    exp = Math.floor(exp / 2)
  }
  return res
}

export { matMul, matPow, matPow as matQPow }

if (require.main === module) {
  // 790. 多米诺和托米诺平铺
  // https://leetcode.cn/problems/domino-and-tromino-tiling/
  function numTilings(n: number): number {
    const MOD = 1e9 + 7
    const init = [[5], [2], [1], [0]]
    if (n <= 3) return init[init.length - 1 - n][0]

    const T = [
      [2, 0, 1, 0],
      [1, 0, 0, 0],
      [0, 1, 0, 0],
      [0, 0, 1, 0]
    ]
    const resT = matPow(T, n - 3, MOD)
    const res = matMul(resT, init, MOD)
    return res[0][0]
  }
}

File: 19_数学/矩阵运算/矩阵快速幂/matqpow.java
// java矩阵快速幂

import java.util.ArrayList;

/**
 * 矩阵快速幂
 */
class MatPow {
  public static long[][] mul(long[][] mat1, long[][] mat2, int mod) {
    long[][] res = new long[mat1.length][mat2[0].length];
    for (int i = 0; i < mat1.length; i++) {
      for (int k = 0; k < mat2.length; k++) {
        for (int j = 0; j < mat2[0].length; j++) {
          res[i][j] = (res[i][j] + mat1[i][k] * mat2[k][j]) % mod;
          if (res[i][j] < 0) {
            res[i][j] += mod;
          }
        }
      }
    }
    return res;
  }

  private int n;
  private int mod;
  private long[] base;
  private int cacheLevel;
  private boolean useCache;
  private ArrayList<ArrayList<long[]>> cache;

  /**
   * 
   * @param base       转移矩阵,必须是方阵
   * @param mod        矩阵快速幂的模数
   * @param cacheLevel
   *                   矩阵快速幂的log底数.启用缓存时一般设置为 {@code 4}.
   *                   当调用 {@link MatPow#pow} 次数很多时,可以设置为{@code 16}.
   *                   小于 {@code 2} 时不启用缓存.
   */
  public MatPow(long[][] base, int mod, int cacheLevel) {
    this.n = base.length;
    this.mod = mod;
    this.base = new long[n * n];
    for (int i = 0; i < n; i++) {
      for (int j = 0; j < n; j++) {
        this.base[i * n + j] = base[i][j];
      }
    }
    this.cacheLevel = cacheLevel;
    this.useCache = cacheLevel >= 2;
    if (useCache) {
      this.cache = new ArrayList<ArrayList<long[]>>(cacheLevel - 1);
      for (int i = 0; i < cacheLevel - 1; i++) {
        this.cache.add(new ArrayList<long[]>(64));
      }
    }
  }

  /**
   * 时间复杂度 {@code O(n^3*logk)}.
   * 
   * @param exp 幂次/矩阵转移次数
   */
  public long[][] pow(long exp) {
    if (!useCache) {
      return powWithoutCache(exp);
    }

    if (cache.get(0).isEmpty()) {
      cache.get(0).add(base);
      for (int i = 1; i < cacheLevel - 1; i++) {
        cache.get(i).add(mul(cache.get(i - 1).get(0), base));
      }
    }

    long[] e = eye(n);
    int div = 0;
    while (exp > 0) {
      if (div == cache.get(0).size()) {
        cache.get(0).add(mul(cache.get(cacheLevel - 2).get(div - 1), cache.get(0).get(div - 1)));
        for (int i = 1; i < cacheLevel - 1; i++) {
          cache.get(i).add(mul(cache.get(i - 1).get(div), cache.get(0).get(div)));
        }
      }

      int mod = (int) (exp % cacheLevel);
      if (mod > 0) {
        e = mul(e, cache.get(mod - 1).get(div));
      }
      exp /= cacheLevel;
      div++;
    }

    return to2D(e);
  }

  private long[] mul(long[] mat1, long[] mat2) {
    long res[] = new long[n * n];
    for (int i = 0; i < n; i++) {
      for (int k = 0; k < n; k++) {
        for (int j = 0; j < n; j++) {
          res[i * n + j] = (res[i * n + j] + mat1[i * n + k] * mat2[k * n + j]) % mod;
          if (res[i * n + j] < 0) {
            res[i * n + j] += mod;
          }
        }
      }
    }
    return res;
  }

  private long[][] powWithoutCache(long exp) {
    long[] e = eye(n);
    long[] b = base.clone();
    while (exp > 0) {
      if ((exp & 1) == 1) {
        e = mul(e, b);
      }
      b = mul(b, b);
      exp >>= 1;
    }

    return to2D(e);
  }

  private long[] eye(int n) {
    long[] res = new long[n * n];
    for (int i = 0; i < n; i++) {
      res[i * n + i] = 1;
    }
    return res;
  }

  private long[][] to2D(long[] mat) {
    long[][] res = new long[n][n];
    for (int i = 0; i < n; i++) {
      for (int j = 0; j < n; j++) {
        res[i][j] = mat[i * n + j];
      }
    }
    return res;
  }
}